ENHANCED: Read more characters at once, significantly speeding up library(pio)

triska opened the pull request:
Enjoy!

notoria commented:
Why read one byte at a time? There is [take](https://doc.rust-lang.org/std/io/trait.Read.html#method.take). And for the name `string` may be matter than `str` (less confusion with the type `str`).

triska commented:
Yes, I noticed `take`, however I found the description unclear in case of errors ("may succeed": what is that supposed to mean, that it may also fail even if the error is resolved?).

In fact I chose `str` also to avoid confusion with the type `String`!

notoria commented:
There is the type `str` used as `&str` or `Box<str>` but for `string` the isn't such type but it's close to `String`.

The current implementation is like `take`, on the first error it halts but that doesn't mean EOF has been reached.

triska commented:
Thank you @notoria for these suggestions, I have addressed these issues in additional commits.

Please review, and merge if applicable! Thank you again.

notoria commented:
I don't have the numbers but an iterator could be faster (and nicer?). Something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=deb8745b1d29ae060effd01ae85ccc52), it could be done for both binary and text mode. With `push` there could be reallocation.

triska commented:
I suggest to leave micro-optimizations to contributors who want to measure the performance of such changes, and ideally address this in Rust itself where possible.

triska commented:
I have rebased the changes against the current master branch!

One additional note: The next and final big step to improve efficiency is to map files without `0`-bytes directly into memory using `mmap` (see #251 ). However, for now, reading in larger chunks as implemented in these commits already improves performance significantly, and also directly uses the compact internal representation for strings.
